docs/howto_contribute.md

```diff
@@ -0,0 +1,396 @@
+symlinks to the top-level `callback_plugins` and `roles` directories.  This
+allows the test to reference the available roles and have its output nicely
+formatted by the callback plugin.  In this example, we are going to pretend we
+are creating a test playbook for the `docker build` command.
+
+```bash
+$ mkdir -p tests/docker-build/
+$ ln -s ../../callback_plugins/ tests/docker-build/callback_plugins
+$ ln -s ../../roles/ tests/docker-build/roles
+```
+
+Additional directories such as `files`, `templates` or `vars` may be required
+for your test.  Please reference the [Ansible documentation](http://docs.ansible.com/ansible/playbooks_best_practices.html#content-organization) about
+content organization about when and how to use these directories.
+
+#### Playbook Structure + System State
+
+Now we can start developing the actual playbook.  When running an Ansible
+playbook, the operations are mostly executed in serial, so there is a sense
+of state of the host under test.  This means you can structure your playbook
+to assume certain conditions of the host during the execution of the playbook.
+
+A simple playbook will have a list of tasks or roles defined in a single file
+and that is all that will be required.  Sometimes, it is desirable to create
+multiple playbooks in a single file for the purposes of providing delineation
+between certain features being tested.  This can be seen in [tests/admin-unlock](/tests/admin-unlock/main.yml)
+where each set of features is separated into their own playbooks.
+
+For our example, we will use the simple case of a single playbook in a single
+file.
+
+#### Playbook Sections
+
+All of the playbooks should start with a `name`, `hosts`, and `become`
+declaration.  The name is up to the test developer, but we almost always use
+`hosts: all` and `become: yes`.  The `hosts: all` allows the executor of the
+test some flexibility in how they define the inventory passed to the
+playbook.  And since most of our test operations require root privileges,
+the `become: yes` declaration ensures that we have those privileges.
+
+The playbook should be tagged via a `tag` value.  This allows the test
+executor the ability to skip the entire playbook, if they were to try to run
+multiple playbooks via a script or other framework.
+
+If the playbook requires additional variables to be defined, they can be
+specified via the `vars_files` or `vars` declaration.
+
+Often a tests will require some tasks to be performed that are not covered
+by any existing roles.  In this situation, one would use the `pre_tasks`
+statement to define the tasks that should be executed ahead of the roles.
+
+After any required `pre_tasks` the test should be built using the roles
+available in the top-level directory.
+
+Finally, any tasks that need to be run after the roles can be specified
+in the `post_tasks` section.
+
+#### Example Playbook
+
+In the example playbook, we are going to copy a couple of Dockerfiles to the
```

Comment:
I would suggest inserting a step #0 in here:  Write a README.md file in the new tests directory :D  The effect is, it makes the author consider the big-picture of what they're trying to accomplish before writing any code.  That almost always results in better quality code AND documentation.  Otherwise the docs. are always an afterthought (or omitted all together).